### PR TITLE
tests - exact exception matching

### PR DESCRIPTION
**Pull request by ttodua**

### CODE REVIEW

**Comment by ttodua** [comment]
@carlosmiei tests passed (unrelated failure)

**Comment by carlosmiei** [comment]
@ttodua Actually I think this defeats the whole point of having hierarchical exceptions, why do we need this? The order of comparisons should be enough, so let's say you have

```
exception A{}
exception B extends A {}
``

If you want to match B and A and treat the errors differently, you just need to check if the exception is B first

**Comment by ttodua** [comment]
but why it defeats..? the test is specifically curated for catching the exact exceptions that are under our "tolerated" list for retries. for other people, the hierarchy could be used for other means, not exactly like this.
i steongly think that relying logic on the order of comparison is the most undesirable and fragile way to code things.  here we need absolutely exact kind of exceptions and it should definitely be agnostic of hierarchy (as typically the lower layer can change the hierarchy anytime, and this snippet shouldnt matter the order).

**Comment by ttodua** [comment]
but as you'll tell.. i'll go and do as you commented

**Comment by carlosmiei** [comment]
@ttodua 

> y (as another layer can change the hierarchy anytime, and one can become another's subclass or in reverse, or even migrate toward anotther ancesto

No that's against the point of having a hierarchy, the goal is to always have the generic exceptions in the base and become more specific on each level, comparing it exactly seems a bit weird and I don't think that is the recommended way of handling it

**Comment by ttodua** [comment]
so, what you say is that we should check if exception is NetworkError, but not ExchangeNotAvailable, ( EthernetJackRemoved, WifiModemShutDown, ProviderIssues... or any subclass that might theoretically be inherited from NetworkError now or in future) and list them all i  this snippet just because to explicitly skip them? and when we add new subclass of NetworkError, we should update this snippet too in order to add it in skips...
well, i obviously see in opposite, but sure, i'm pushing it now as you say.

**Comment by carlosmiei** [comment]
@ttodua sorry I'm not understading the point here, for instance right now we do this:

```
(...)
const isRateLimitExceeded = (e instanceof RateLimitExceeded);
```

this will only be true when the exception is a RateLimitExceeded exception or something more specific than that, right? so what's the problem? if it's something more specific like RateLimitExceededPerUser that's also an RateLimitExceeded error

**Comment by ttodua** [comment]
> @ttodua sorry I'm not understading the point here, for instance right now we do this:
> 
> ```
> (...)
> const isRateLimitExceeded = (e instanceof RateLimitExceeded);
> ```
> 
> this will only be true when the exception is a RateLimitExceeded exception or something more specific than that, right? so what's the problem? if it's something more specific like RateLimitExceededPerUser that's also an RateLimitExceeded error

so, what I mean is that `ExchangeNotAvailable` is subclass of `NetworkError`. typically we tolerate `NetworkError` kind to schedule it for retry and even in case of repeated failure, we still "pass" the exchange, because, in `NetworkError` we blame somewhat interminent temporary failure, that does not directly point to our implementational issues. 
however, specifically `ExchangeNotAvailable` is not tolerated for retry, because, it mean either API is changed or something important issue, that should not be skipped and we should throw exception. that's why, we specifically need the generic/vague `NetworkError` exception, exactly that (not its subclasses) to consider that "hmm, something unspecified network error, let's skip it and mark it passed for now, because it neither points to any changes in API engine, neither our implementation issue, so let's dont break the build"

**Comment by carlosmiei** [comment]
@ttodua in this case you can catch the exceptions using different try/catch following the intended order, eg:

```Python
   except ccxt.ExchangeNotAvailable as e:
      pass
   except ccxt.Network as e:
      pass
   except Exception as e:
      pass
```

**Comment by ttodua** [comment]
that's I was saying/opposing, as we had to list exceptions (whereas those exceptions shouldn't be needed to be mentioned in tests itself, and therefore if we expand NetworkError subclass in future, we had to add new "skips" here, and thus we become dependent to "skips", and also rely on the order of if/else, while we could have just directly compared the type and avoided those burden. 

but oky, in 1 min i'll push the way you say.

**Comment by ttodua** [comment]
i'm digging into it and seems there are diversified inheritance,  `ExchangeNotAvailable` being subclass of `NetworkError`, but `OnMaintenance` seems to be subclass of `ExchangeNotAvailable`. and other exceptions too, nested with one another's inheritance. jfyi, that will the most fragile if/else what we will be doing now.

**Comment by carlosmiei** [comment]
> jfyi, i'm digging into it and seems there are diversified inheritance, `ExchangeNotAvailable` (not tolerable for skip) being subclass of `NetworkError` (tolerable for skip) , but `ExchangeNotAvailable` (not tolerable for skip) is ancestor of `OnMaintenance` (tolerable for skip). and other exceptions too, nested with one another's inheritance. that will the most fragile if/else chain what we will be doing now.

@ttodua Maybe we have to change our requirements then, does not make sense to strip the hierarchy of the exceptions, it goes against its concept. Second point, I don't know why following an order is fragile they're arranged from the most basic exception to the most specific one.

**Comment by ttodua** [comment]
so, what I mean, we make retry on specifically these exceptions: `NetworkError, RateLimitExceeded, DDoSProtection, RequestTimeout`.
oky, I've pushed your version, please take a look if that's acceptable. You can take see `const tempFailure =` ... line that I meant to be not much intuitive to have negate clauses, as opposed to writing down the exact (target) exceptions we needed specifically. but if this commit & PR looks good, lets go ahead forward merging.

**Comment by ttodua** [comment]
btw, I am not sure why build fails and the error it shows (the unused imports) are actually strange as those imports do not exist at all

File: php/test/test_async.php [hunk 1]
```diff
@@ -135,6 +135,13 @@ function exception_message($exc) {
     return substr($message, 0, LOG_CHARS_LENGTH);
 }
 
+function compare_exception_type($exc, $exceptionType) {
+    // $classFullName = '\\' . get_class($exc);
+    // return $classFullName === $exceptionType; (`$exceptionType` is string here)
+    //
+    // update: for now we check the exception and it's inheritances
+    return ($exc instanceof $exceptionType);
+}
```
Comment (ttodua) [inline php/test/test_async.php:144]: ..

**Comment by ttodua** [comment]
@carlosmiei PR ready

**Comment by ttodua** [comment]
PR ready.

File: python/ccxt/test/test_async.py [hunk 2]
```diff
@@ -161,7 +161,14 @@ def exception_message(exc):
     return message
 
 
-def exit_script(code=0):
+def compare_exception_type(exc, exceptionType):
```
Comment (carlosmiei) [inline python/ccxt/test/test_async.py:164]: @ttodua I would suggest removing this method it adds complexity for no reason, I would rather see the instance of directly inside the catch

Comment (ttodua) [reply python/ccxt/test/test_async.py]: done ;

**Comment by carlosmiei** [comment]
LGTM

File: python/ccxt/test/test_async.py [hunk 3]
```diff
@@ -161,7 +161,7 @@ def exception_message(exc):
     return message
 
 
-def exit_script(code=0):
```
Comment (carlosmiei) [inline python/ccxt/test/test_async.py:164]: @ttodua this is needed now

Comment (carlosmiei) [reply python/ccxt/test/test_async.py:164]: ( I mean the code, we can let the default in case we forget to provide it)

**Pull request merged**